keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,656 @@
+  finalizers:
+    - cosi.io/finalizer [3]
+spec:
+  provisioner: [4]
+  releasePolicy: [5]
+  anonymousAccessMode: [6]
+  bucketClassName: [7]
+  bindings: [8]
+    - "<BucketAccess.name>"
+  protocol: [9]
+    protocolSignature: ""
+    azureBlob:
+      containerName:
+      storageAccount:
+    s3:
+      endpoint:
+      bucketName:
+      region:
+      signatureVersion:
+    gcs:
+      bucketName:
+      privateKeyName:
+      projectId:
+      serviceAccount:
+  allowedNamespaces: [10]
+    - name:
+      uid:
+  parameters: [11]
+status:
+  conditions: [12]
+    - Type: {BucketAvailable} [13]
+      Status: [14]
+      Reason: [15]
+      Message: [16]
+```
+
+1. `name`: When created by COSI, the `Bucket` name is generated in this format: _<bucketRequest.namespace>"-"<bucketRequest.name>_. If an admin creates a `Bucket`, as is necessary for brownfield access, they can use any name.
+2. `labels`: added by the controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+3. `finalizers`: added by the controller to defer `Bucket` deletion until backend deletion ops succeed.
+4. `provisioner`: The provisioner field as defined in the `BucketClass`.  Used by sidecars to filter `Bucket`s. Format: <provisioner-namespace>"/"<provisioner-name>, eg "ceph-rgw-provisoning/ceph-rgw.cosi.ceph.com".
+5. `releasePolicy`: Prescribes outcome of a Delete events. 
+   - _Retain_: (default) the `Bucket` and its data are preserved. The `Bucket` can potentially be reused.
+   - _Delete_: the bucket and its contents are destroyed.
+> Note: the `Bucket`'s release policy is set to "Retain" as a default. Exercise caution when using the "Delete" release policy as the bucket content will be deleted.
+> Note: a `Bucket` is not deleted if it is bound to any `BucketRequest`s.
+6. `anonymousAccessMode`: a string specifying *uncredentialed* access to the Bucket.  This is applicable to cases where the storage instance or objects are intended to be publicly readable and/or writable. One of:
+   - "private": Default, disallow uncredentialed access to the storage instance.
+   - "publicReadOnly": Read only, uncredentialed users can call ListBucket and GetObject.
+   - "publicWriteOnly": Write only, uncredentialed users can only call PutObject.
+   - "publicReadWrite": Read/Write, same as `ro` with the addition of PutObject being allowed.
+> Note: does not reflect or alter the backing storage instances' ACLs or IAM policies.
+7. `bucketClassName`: Name of the associated bucket class (greenfield only).
+8. `bindings`: an array of `BucketAccess.name`(s). If the list is empty then there are no bindings (accessors) to this `Bucket` instance and the `Bucket` can potentially be deleted.
+9. `protocol`: The protocol the application will use to access the storage instance.
```

Comment:
I believe this design makes it impossible for a single `Bucket` (its objects) to be exposed using multiple protocols. Is this intentional? What's the suggested approach for storage systems which support multi-protocol access to the same buckets?

Reply:
There is no object storage provider that provides multiple protocol support for any single bucket. The only way to access a single bucket with different protocols is by using a protocol-translating proxy. eg. S3Proxy

In case the proxy is being utilized, multiple bucket objects can be created for each of the protocols, and have them point to their respective proxies.

Reply:
> There is no object storage provider that provides multiple protocol support for any single bucket. The only way to access a single bucket with different protocols is by using a protocol-translating proxy. eg. S3Proxy

This is incorrect. Our solution allows a single bucket and its object to be accessed using either an AWS S3-compatible API, or the Azure Blob-compatible API, with authn/z interop, without `s3proxy` or something similar. As such, this design limitation directly affects/limits creating large-scale applications where some services use one API, others another, to access the same datastore.